readme first draft (#1)
avohq/avo_audit: README.md

# Avo Audit

Avo audit helps you detect issues in your raw analytics event tables by querying a sample of the raw events and comparing them against each other.
This allows you to find discrepancies between events and properties that should look the same, but don’t.

## Issue types

> Issues available in the alpha version:  

Reviewer: ```suggestion
### Issues available in the alpha version:  
```

**Property sometimes missing** – detects absence of event properties across tracking calls.  

> >`Property onboarding status is not sent with 30% of the App Opened event.`  
> 
> **Property type mismatch** – detects inconsistent event property types across tracking calls.  
> >`Inconsistent type of property onboarding status on event App Opened: int (84%), string (16%).`

Reviewer: ```suggestion
>_Property onboarding status is not sent with 30% of the App Opened event._ 

**Property type mismatch** – detects inconsistent event property types across tracking calls.  
>_Inconsistent type of property onboarding status on event App Opened: int (84%), string (16%)._
```

Reviewer 2: Yeah I agree this looks a bit too heavy on formatting now

### Issue types on the roadmap (not prioritized list):
* Event missing on some platforms
* Property missing on some platforms
* Event volume change significant between versions
* Event volume significantly different between platforms
* Inconsistent event name casing
* Inconsistent property name casing
* Global property type mismatch
* Similar event names
* Similar property names within event
* Unexpected type based on property name
* Global similar property names
* Missing property based on property group pattern

## How do I get the most out of this package? 

> We recommend implementing at least the main 3 macros, and chain them together. That way you will generate a SQL table listing the issues found in your project, ordered by the priority listed above.

Reviewer: ```suggestion
We recommend implementing at least the main 3 macros, and chain them together. That way you will generate a SQL table listing the issues found in your project.
```

Steps:
1. Convert raw events into it’s signature type with **parse_signature**

> 2. Group events by identical signatures and compare events with different signatures Per event with **compare_events**

Reviewer 2: Compare events with different signatures per event name?

> 3. Generate issues table by using the **report_issues** macro.

Reviewer: Should we link to the macros here somehow? Either some sort of a reference doc, or directly to the code (if it exists)

## Installing the package
Include the following in your `packages.yml` file in your dbt project:

```
packages:
    - package: avo/avo_audit
      version: ["0.1.0]
```

Run `dbt deps` in the root of your dbt project to install the package.

# Macros

> ### Disclaimer!

Reviewer: ```suggestion
### Disclaimer
```

> > We have not yet implemented these macros, however some of the work has been started and we will be doing our best to get this up and running as soon as possible!

Reviewer: ```suggestion
> This project is still work in progress. These macros have not been fully implemented yet, however some of the work has been started and we will be doing our best to get this up and running as soon as possible. Reach out if you'd like to help out!
```

> > This project is still work in progress. These macros have not been fully implemented yet, however some of the work has been started and we will be doing our best to get this up and running as soon as possible. Reach out if you'd like to help out!

Reviewer 3: I feel like "Help out" here is a bit vague. Maybe instead of "Reach out if you'd like to help out!" we could say something like "Reach out if you want to join our alpha group to be one of the first to try out what we build and provide feedback!"

## parse_signature

This macro generates a SQL that transforms each raw event into it’s signature.  
Instead of using the signature from the column definitions, it is deriving the signature from the value. This can be used to compare events with identical names together to understand if and how they vary from each other.

**Raw Events:**
| event_name  | user_id      | user_name | device  |
|-------------|--------------|-----------|---------|
| App Opened  | 2c3416a3033e | John      | IOS     |
| App Opened  | fc3416basdq1 | Donna     | IOS     |
| App Opened  | c5e6d9429f27 | Lisa      |         |

> | App Opened  | 55331212412  | Joshua    | Android |

Reviewer 2: 55331212412 is over 32 bits so can be classified as big int or something 🤔

**After running parse_signature:**
| event_name | property_name_mapping        | property_signature_mapping |
|------------|------------------------------|----------------------------|
| App Opened | [user_id, user_name, device] | [string, string, string]   |
| App Opened | [user_id, user_name, device] | [string, string, string]   |
| App Opened | [user_id, user_name, device] | [string, string, null]     |
| App Opened | [user_id, user_name, device] | [int, string, null]        |

> It is best to test this in your develop environment first while you get set up.

Reviewer: ```suggestion
We recommend to test this in your development environment first while you get set up.
```

```
{# in dbt Develop #}

{% set raw_event_relation=adapter.get_relation(
      database=target.database,
      schema="raw_event_schema",
      identifier="raw_event_table"
) -%}

{{ avo_audit.parse_signature(
    relation=raw_event_relation,
    date_partition=”sent_at”,
    lookback_hours=1,
) }}
```

**Arguments:**
* **raw_event_relation:** The database table containing your raw events which you want to compare.
* **date_partition:** The column in raw events that can be used to filter on time to limit query.
* **lookback_hours:** How many hours the query should look back, we recommend keeping it under 24h to keep the cost of the query low.

## compare_events

> This macro will run on the results of parse_signature.

Reviewer 2: So this result is stored in the database somewhere and compare_events works on top of that table?

Author: Yes, each step creates a new table, which the next works on top of. 
I just added an "audit" macro which combines all 3 so teh user does not have to create all 3 models if they don't want to.

> It will execute a Sql that will go through each event_name group, and compare it together with all the other events to find issues where a field is sometimes NULL, Int instead of string etc.

Reviewer: ```suggestion
It will execute a SQL that will go through each event_name group, and compare it together with all the other events to identify the issues documented above.
```

> It will execute a SQL that will go through each event_name group, and compare it together with all the other events to identify the issues documented above.

Reviewer 2: I'm not following how this works 🤔

Author: I updated this description, is it any clearer now ?

Reviewer 2: Yes, really clear now!

> It will execute a SQL that will go through each event_name group, and compare it together with all the other events to identify the issues documented above.

Reviewer 2: ```suggestion
It will go through each event_name group, and compare it together with all the other events to identify the missing properties, property type mismatches and other issues documented above.
```

Reviewer 2: What is exactly an event_name group?

The end result will be a table that identifies discrepancies between the events.

| event_name | property_name_mapping        | property_signature_mapping | variant_count | total_event_count | number_of_event_variants |
|------------|------------------------------|----------------------------|---------------|-------------------|--------------------------|
| App Opened | [user_id, user_name, device] | [string, string, string]   | 2             | 4                 | 3                        |
| App Opened | [user_id, user_name, device] | [string, string, null]     | 1             | 4                 | 3                        |
| App Opened | [user_id, user_name, device] | [int, string, null]        | 1             | 4                 | 3                        |

> 

Reviewer 3: Add some context for the commands here? Like "To execute, run the following commands in your dbt dev environment:"

```
{# in dbt Develop #}

{% set raw_event_signature_relation=adapter.get_relation(
      database=target.database,
      schema="avo_audit_schema",
      identifier="raw_event_signature_table"
) -%}

{{ avo_audit.parse_signature(
    relation=raw_event_signature_relation
) }}
```

## report_issues

This macro takes the result from compare_events, and generates issues based on number of event variants, which kind of variant difference it is etc. The issues are priorities in the following order.

* **Type difference** - Inconsistent type of property onboarding status on event App Opened: int (84%), string (16%).
* **Wrong casing** - property name is not following the general casing of the project.

> * **Property Sometimes missing** -  Property onboarding status is not sent with 30% of the App Opened event.

Reviewer: Update based on updated issues documentation above?

| event_name | property_name_mapping        | property_signature_mapping | variant_count | total_event_count | number_of_event_variants | issue_priority | issue                                                                                 |
|------------|------------------------------|----------------------------|---------------|-------------------|--------------------------|----------------|---------------------------------------------------------------------------------------|
| App Opened | [user_id, user_name, device] | [int, string, null]        | 1             | 4                 | 3                        | 1              | Inconsistent type of property "user_id" on event App Opened: int (25%), string (75%). |
| App Opened | [user_id, user_name, device] | [string, string, null]     | 1             | 4                 | 3                        | 4              | Property "device" is not sent with 25% of the App Opened event.                       |

> **Arguments:**

Author: Added

* **compare_event_relation:** The database table containing your raw events which you want to compare.
* **report_missing_property:** Whether the report should include property missing, defaults to true

# Upcoming Macros

> ### Disclaimer!
> > Here we list macro's that we want to work on in the future and have been mapped out how they should work. If you want to contribute, this list of macros is up for grabs!

Reviewer: ```suggestion
### Disclaimer
> Here we list macro's that we plan on building. If you want to contribute, this list of macros is up for grabs!
```

## audit_volume

> This macro is intended to compare data from two time periods to each other, and compare for each event relative to number of events received to identify if any event or event signature significantly drops or rises in volume.

Reviewer 2: ```suggestion
This macro is intended to compare data from two time periods to each other and find significant drops or rises in event or event signature volumes.
```

> This macro is intended to compare 2 days of data to each other, and compare for each event relative to number of events received to identify if any event or event signature significantly drops or rises in volume.

Reviewer: the "2 days" part would not necessarily be fixed right?
```suggestion
This macro is intended to compare data from two time periods to each other, and compare for each event relative to number of events received to identify if any event or event signature significantly drops or rises in volume.
```

**Property sometimes missing** – detects absence of event properties across tracking calls.  
>_Property onboarding status is not sent with 30% of the App Opened event._ 

**Property type mismatch** – detects inconsistent event property types across tracking calls.  
>_Inconsistent type of property onboarding status on event App Opened: int (84%), string (16%)._

### Issue types on the roadmap (not prioritized list):
* Event missing on some platforms
* Property missing on some platforms
* Event volume change significant between versions
* Event volume significantly different between platforms
* Inconsistent event name casing
* Inconsistent property name casing
* Global property type mismatch
* Similar event names
* Similar property names within event
* Unexpected type based on property name
* Global similar property names
* Missing property based on property group pattern

## How do I get the most out of this package? 
We recommend implementing at least the main 3 macros, and chain them together. That way you will generate a SQL table listing the issues found in your project.

Steps:
1. Convert raw events into it’s signature type with [**parse_signature**](#parse_signature)

> 2. Group events by identical signatures and compare events with different signatures Per event with [**compare_events**](#compare_events)

Reviewer 2: Maybe just
```suggestion
2. Group events by identical signatures and compare events with different signatures with [**compare_events**](#compare_events)
```

This allows you to find discrepancies between events and properties that should look the same, but don’t.

## Issue types

### Issues available in the alpha version:  

**Property sometimes missing** – detects absence of event properties across tracking calls.  
>_Property onboarding status is not sent with 30% of the App Opened event._ 

**Property type mismatch** – detects inconsistent event property types across tracking calls.  
>_Inconsistent type of property onboarding status on event App Opened: int (84%), string (16%)._

### Issue types on the roadmap (not prioritized list):
* Event missing on some platforms
* Property missing on some platforms
* Event volume change significant between versions
* Event volume significantly different between platforms
* Inconsistent event name casing
* Inconsistent property name casing
* Global property type mismatch
* Similar event names
* Similar property names within event
* Unexpected type based on property name
* Global similar property names
* Missing property based on property group pattern

## How do I get the most out of this package? 

> We recommend implementing at least the main 3 macros, and chain them together. That way you will generate a SQL table listing the issues found in your project.

Reviewer 2: Or just use `audit`?
